src/nvim/api/buffer.c

```diff
@@ -722,8 +722,16 @@ void nvim_buf_set_text(uint64_t channel_id, Buffer buffer, Integer start_row, In
   FOR_ALL_TAB_WINDOWS(tp, win) {
     if (win->w_buffer == buf) {
       // adjust cursor like an extmark ( i e it was inside last_part_len)
-      if (win->w_cursor.lnum == end_row && win->w_cursor.col > end_col) {
-        win->w_cursor.col -= col_extent - (colnr_T)last_item.size;
+      if (win->w_cursor.lnum == end_row) {
+        if (win == curwin && (State & MODE_INSERT)) {
```

Comment:
I think this makes sense. in extmark terms, the insert mode cursor has different "gravity" than normal (and non-current win) cursor. 

however instead of code duplication I suggest write some thing like `int insert_adjust = (win == curwin && (State & MODE_INSERT)) ? 1 : 0` and then use that adjustment in a non-duplicated if statement.